src/iterators/intersection_iterator.c

```diff
@@ -308,6 +308,50 @@ static QueryIterator *IntersectionIteratorReducer(QueryIterator **its, size_t *n
   return ret;
 }
 
+static ValidateStatus II_Revalidate(QueryIterator *base) {
+  IntersectionIterator *ii = (IntersectionIterator *)base;
+  bool any_child_moved = false, movedToEOF = false;
+  t_docId max_child_docId = 0;
+
+  // Step 1: Revalidate all children and track status
+  for (uint32_t i = 0; i < ii->num_its; i++) {
+    QueryIterator *child = ii->its[i];
+    ValidateStatus child_status = child->Revalidate(child);
+
+    if (child_status == VALIDATE_ABORTED) {
+      return VALIDATE_ABORTED; // Intersection fails if any child fails
+    }
+
+    if (child_status == VALIDATE_MOVED) {
+      any_child_moved = true;
+      // Track the maximum docId among moved children
+      if (child->atEOF) {
+        movedToEOF = true; // If any child moved and now at EOF, the intersection is also at EOF now
+      } else if (child->lastDocId > max_child_docId) {
+        max_child_docId = child->lastDocId;
+      }
+    }
+  }
+
+  // Step 2: Handle the result based on child status
+  if (!any_child_moved) {
+    // All children returned OK - simply return OK
+    return VALIDATE_OK;
+  } else if (base->atEOF) {
```

Comment:
If it is already at EOF, is it needed to do all the previous revalidation on all the children?

Reply:
We should revalidate in case we get rewound. I think we should take the safe approach for now, but if we ever see it will be beneficial to revalidate only if needed, we can add a ”needs revalidation” flag and revalidate on rewind